### DISCUSSION TITLE
An issue about C++ grammer in moose

### DISCUSSION THREAD

**Discussion by xiaojbing** [idaholab/moose#20889] [Q&A General] [answered]
I add code in my application to solve CFD problems. And I find that different conditional  statement will cause different residual and results in the following code. This code is to compute Jacobi Matrix and inherits DGkernel. The difference of conditional statement is just at orange area. But this should not cause difference,right?
conditional  statement 1:
![image](https://user-images.githubusercontent.com/87591264/165507895-07f0df69-1e5d-473a-9014-6646eac9a3e5.png)
residual and results:
![image](https://user-images.githubusercontent.com/87591264/165508006-2c8ac1b7-43f5-488f-a181-e6c10cc29d38.png)
![image](https://user-images.githubusercontent.com/87591264/165508028-ab3829ec-e73e-4ddb-958d-7e3aa3300b8d.png)

conditional  statement 2:
![image](https://user-images.githubusercontent.com/87591264/165508419-85ba0d7e-13a5-45cb-b06e-e0d2c11773ca.png)
residual and results:
![image](https://user-images.githubusercontent.com/87591264/165508484-1412335b-7b40-40d2-a325-ee63d5d5300d.png)
![image](https://user-images.githubusercontent.com/87591264/165508512-5ea544d6-eeb7-4756-bb71-db2d98241c04.png)

**Comment by cpgr** [comment]
Without looking at the code at all, I suspect that you just haven't converged as either your relative tolerance  is too loose, or your absolute tolerance is too high. The residual has only dropped by one order of magnitude. Maybe check that first?

**Reply by xiaojbing** [reply]
> Without looking at the code at all, I suspect that you just haven't converged as either your relative tolerance is too loose, or your absolute tolerance is too high. The residual has only dropped by one order of magnitude. Maybe check that first?

Thank you. This code inherit from kernel.C. But the different type conditional statement should not affect residual,right.

**Reply by cpgr** [reply]
No idea! I'll leave that to one of the MOOSE team.

**Comment by loganharbour** [comment, marked as answer]
First - please refrain from posting screenshots of text. It can be difficult to read and is also difficult for other users when searching questions in the discussions in the future.

Second - I would strongly suggest that you do not try to re-implement basic DG capability. In general, you should override `computeQpResidual(Moose::DGResidualType type)` and `computeQpJacobian(Moose::DGJacobianType type)`. Lower level code like this is subject to change.

Now... in terms of the rest of your confusion. This is expected. When you first assign `Kxx` in either case, you're setting to to either the block Jacobian or the neighbor block Jacobian. In what follows (your conditional statements), you are not _reassigning_ `Kxx` but  you are instead calling the equals operator, thus you are _setting_ the Jacobian that you had previously defined. That is:

```c++
DenseMatrix<Number> & Kxx = _assembly.jacobianBlockNeighbor(type, p, q);
// Sets the neighbor Jacobian = to the non-neighbor jacobian
if (type == Moose::ElementElement)
    Kxx = _assembly.jacobianBlock(p, q);
// Sets the neighbor Jacobian = to the neighbor Jacobian
else
    Kxx = _assembly.jacobianBlockNeighbor(type, p, q);
```

You are not re-assigning `Kxx`. You are literally setting `Kxx` equal to some other matrix.
